File: ios/Betty/Features/GroupManagement/JoinInviteSheet.swift
import SwiftUI

/// Outcome of `POST /join/:code` (web `JoinGroupModal.join`), mapped from the API error.
/// 409 = already a member (offer navigation), 404 = invalid/expired invite,
/// 403 = blocked; anything else is the generic failure.
nonisolated enum JoinInviteOutcome: Equatable, Sendable {
    case joined(groupID: Int)
    case alreadyMember
    case invalidInvite
    case blocked
    case failed

    static func map(_ error: any Error) -> JoinInviteOutcome {
        guard let apiError = error as? APIError else { return .failed }
        switch apiError.status {
        case 409: return .alreadyMember
        case 404: return .invalidInvite
        case 403: return .blocked
        default: return .failed
        }
    }
}

/// Web `/dashboard/groups/join/[code]` + `JoinGroupModal` — the invite deep-link entry.
///
/// Loads the preview via `GET /group/:code`, then joins via `POST /join/:code`.
/// Presented as a sheet by the deep-link router (`betty://join/<code>` and the
/// universal link `https://betty.social/dashboard/groups/join/<code>`).
struct JoinInviteSheet: View {
    let code: String

    @Environment(AppEnvironment.self) private var env
    @Environment(ThemeStore.self) private var theme
    @Environment(\.dismiss) private var dismiss

    private enum Phase {
        case loading
        case loaded(GroupPeek)
        case failed
    }

    @State private var phase: Phase = .loading
    @State private var isJoining = false

    var body: some View {
        ZStack {
            theme.colors.surface.ignoresSafeArea()
            switch phase {
            case .loading:
                ProgressView()
                    .tint(theme.colors.textPrimary)
            case .loaded(let peek):
                invite(peek)
            case .failed:
                loadError
            }
        }
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
        .task {
            do {
                phase = .loaded(try await env.groupStore.peek(code: code))
            } catch {
                phase = .failed
            }
        }
    }

    // MARK: invite preview (web JoinGroupModal)

    private func invite(_ peek: GroupPeek) -> some View {
        VStack(spacing: 0) {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    if let headerImageURL = peek.headerImageURL, let url = URL(string: headerImageURL) {
                        heroImage(url: url, tournamentImageURL: peek.tournamentImageURL)
                    }
                    VStack(alignment: .leading, spacing: Space.xs) {
                        if peek.headerImageURL == nil,
                           let tournamentImageURL = peek.tournamentImageURL,
                           let url = URL(string: tournamentImageURL) {
                            roundLogo(url: url, size: 96)
                                .padding(.bottom, Space.s)
                        }
                        Text("★ INVITED TO BET")
                            .kicker(Palette.orange)
                        Text(peek.name.uppercased())
                            .font(.bettyDisplayL)
                            .displayKerning(40)
                            .foregroundStyle(theme.colors.textPrimary)
                            .fixedSize(horizontal: false, vertical: true)
                            .accessibilityIdentifier("joinInvite.invite.groupName")
                        if !peek.tournamentName.isEmpty {
                            Text(peek.tournamentName.uppercased())
                                .font(.bettySubhead)
                                .kerning(1.2)
                                .foregroundStyle(theme.colors.textSecondary)
                        }
                        // Web JoinGroupModal renders the group description behind an
                        // orange left border.
                        if let description = peek.description, !description.isEmpty {
                            BettyInsetPanel(accent: Palette.orange) {
                                Text(description)
                                    .font(.betty(14, .regular))
                                    .foregroundStyle(theme.colors.textSecondary)
                                    .lineSpacing(3)
                            }
                            .padding(.top, Space.xs)
                        }
                        Text("Lock in your bets every matchday, climb the standings, settle the banter.")
                            .font(.betty(14, .regular))
                            .foregroundStyle(theme.colors.textSecondary)
                            .lineSpacing(3)
                            .padding(.top, Space.xs)
                    }
                    .padding(Space.l)
                    .padding(.top, peek.headerImageURL == nil ? Space.s : Space.m)
                }
            }

            Divider()
                .overlay(theme.colors.overlay06)

            HStack(spacing: 0) {
                Button("NO THANKS") {
                    dismiss()
                }
                .buttonStyle(.bettyGhost)
                .frame(maxWidth: .infinity)
                .accessibilityIdentifier("joinInvite.invite.declineButton")

                Button(isJoining ? "PLACING…" : "I'M IN →") {
                    join(peek)
                }
                .buttonStyle(.bettyPrimaryBlock)
                .disabled(isJoining)
                .frame(maxWidth: .infinity)
                .accessibilityIdentifier("joinInvite.invite.acceptButton")
            }
            .padding(Space.m)
        }
    }

    private func heroImage(url: URL, tournamentImageURL: String?) -> some View {
        AsyncImage(url: url) { phase in
            if let image = phase.image {
                image.resizable().scaledToFill()
            } else {
                theme.colors.overlay04
            }
        }
        .aspectRatio(16 / 9, contentMode: .fit)
        .clipped()
        .overlay(alignment: .bottomLeading) {
            if let tournamentImageURL, let iconURL = URL(string: tournamentImageURL) {
                roundLogo(url: iconURL, size: 56)
                    .padding(.leading, Space.l)
                    .offset(y: 22)
            }
        }
    }

    private func roundLogo(url: URL, size: CGFloat) -> some View {
        AsyncImage(url: url) { phase in
            if let image = phase.image {
                image.resizable().scaledToFill()
            } else {
                theme.colors.overlay06
            }
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay {
            Circle().strokeBorder(theme.colors.overlay10, lineWidth: 2)
        }
        .background {
            Circle().fill(theme.colors.surface)
        }
    }

    // MARK: load failure (web join/[code].vue error state)

    private var loadError: some View {
        VStack(alignment: .leading, spacing: Space.m) {
            Text("★ INVITED TO BET")
                .kicker(Palette.orange)
            Text("COULD NOT LOAD THIS INVITE")
                .font(.bettyTitle2)
                .foregroundStyle(theme.colors.textPrimary)
                .fixedSize(horizontal: false, vertical: true)
                .accessibilityIdentifier("joinInvite.error.title")
            Text("The invite link may be invalid or expired. Please check the link and try again.")
                .font(.betty(14, .regular))
                .foregroundStyle(theme.colors.textSecondary)
                .lineSpacing(3)
            Button("GO TO DASHBOARD") {
                dismiss()
                env.router.selectedTab = .home
            }
            .buttonStyle(.bettyOutline)
            .accessibilityIdentifier("joinInvite.error.dashboardButton")
            Spacer()
        }
        .padding(Space.l)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
    }

    // MARK: join

    private func join(_ peek: GroupPeek) {
        isJoining = true
        Task {
            defer { isJoining = false }
            let outcome: JoinInviteOutcome
            do {
                outcome = .joined(groupID: try await env.groupStore.join(code: code))
            } catch {
                outcome = JoinInviteOutcome.map(error)
            }
            handle(outcome, peek: peek)
        }
    }

    private func handle(_ outcome: JoinInviteOutcome, peek: GroupPeek) {
        switch outcome {
        case .joined(let groupID):
            dismiss()
            env.toasts.confirm(question: "You are now a proud member of \(peek.name). Go there now?") {
                navigate(to: groupID)
            }
        case .alreadyMember:
            dismiss()
            env.toasts.confirm(question: "It looks like you're already member of \(peek.name). Go there now?") {
                navigate(to: peek.id)
            }
        case .invalidInvite:
            env.toasts.alert(
                title: "Could not join group",
                message: "This invite link is invalid or has expired.",
                state: .critical
            )
        case .blocked:
            env.toasts.alert(
                title: "Cannot bet here",
                message: "You have been blocked from \(peek.name).",
                state: .warning
            )
        case .failed:
            env.toasts.alert(
                title: "Could not join group",
                message: "Something went wrong while joining the group. Please try again.",
                state: .critical
            )
        }
    }

    private func navigate(to groupID: Int) {
        env.router.selectedTab = .home
        env.router.homePath = [.groupDetail(groupID: groupID)]
    }
}

File: android/app/src/main/java/social/betty/features/join/JoinInviteSheet.kt
package social.betty.features.join

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Divider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import kotlinx.coroutines.launch
import social.betty.core.model.GroupPeek
import social.betty.core.net.ApiError
import social.betty.designsystem.BettyTheme
import social.betty.designsystem.Palette
import social.betty.designsystem.Space
import social.betty.designsystem.components.BettyButton
import social.betty.designsystem.components.BettyButtonVariant
import social.betty.designsystem.components.KickerText
import social.betty.navigation.LocalAppContainer
import social.betty.navigation.LocalNavigator

// ---------------------------------------------------------------------------
// Local phase + outcome types
// ---------------------------------------------------------------------------

private sealed interface Phase {
    data object Loading : Phase
    data class Preview(val peek: GroupPeek) : Phase
    data object PreviewError : Phase
}

/** Result of `POST /join/:code`, mirroring the iOS `JoinInviteOutcome`. */
private sealed interface JoinOutcome {
    data class Joined(val groupId: Int) : JoinOutcome
    data class AlreadyMember(val groupId: Int) : JoinOutcome
    data object InvalidInvite : JoinOutcome
    data object Blocked : JoinOutcome
    data object Failed : JoinOutcome
}

private fun mapJoinError(e: Throwable, fallbackGroupId: Int): JoinOutcome =
    when ((e as? ApiError)?.statusCode) {
        409 -> JoinOutcome.AlreadyMember(fallbackGroupId)
        404 -> JoinOutcome.InvalidInvite
        403 -> JoinOutcome.Blocked
        else -> JoinOutcome.Failed
    }

/**
 * The invite deep-link landing sheet (screens.md §3.6, components.md §4.3).
 *
 * Loads the group preview via GET /group/:code, then joins via POST /join/:code.
 * Called by the shell's SheetHost when Sheet.JoinInvite(code) is active.
 *
 * Column content only — the ModalBottomSheet wrapper is provided by SheetHost.
 */
@Composable
fun JoinInviteSheet(code: String, onDismiss: () -> Unit) {
    val container = LocalAppContainer.current
    val nav = LocalNavigator.current
    val scope = rememberCoroutineScope()

    var phase by remember { mutableStateOf<Phase>(Phase.Loading) }
    var isJoining by remember { mutableStateOf(false) }

    // Pending confirm dialog: holds the question text + the action to run on "Go there now?"
    var pendingConfirm by remember { mutableStateOf<Pair<String, () -> Unit>?>(null) }

    LaunchedEffect(code) {
        phase = try {
            Phase.Preview(container.api.getGroupByCode(code))
        } catch (_: Exception) {
            Phase.PreviewError
        }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .testTag("join-invite-sheet"),
    ) {
        when (val p = phase) {
            Phase.Loading -> {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = Space.xxl),
                    contentAlignment = Alignment.Center,
                ) {
                    CircularProgressIndicator(
                        color = BettyTheme.colors.textPrimary,
                        modifier = Modifier.size(36.dp),
                    )
                }
            }

            is Phase.Preview -> {
                InvitePreview(
                    peek = p.peek,
                    isJoining = isJoining,
                    onDecline = onDismiss,
                    onAccept = {
                        isJoining = true
                        scope.launch {
                            val outcome = try {
                                val groupId = container.groupStore.joinByCode(code)
                                JoinOutcome.Joined(groupId)
                            } catch (e: Exception) {
                                mapJoinError(e, p.peek.id)
                            }
                            isJoining = false
                            when (outcome) {
                                is JoinOutcome.Joined -> {
                                    onDismiss()
                                    pendingConfirm = Pair(
                                        "You are now a proud member of ${p.peek.name}. Go there now?",
                                        { nav.openGroup(outcome.groupId) },
                                    )
                                }
                                is JoinOutcome.AlreadyMember -> {
                                    onDismiss()
                                    pendingConfirm = Pair(
                                        "It looks like you're already member of ${p.peek.name}. Go there now?",
                                        { nav.openGroup(outcome.groupId) },
                                    )
                                }
                                JoinOutcome.InvalidInvite ->
                                    container.notify.error("This invite link is invalid or has expired.")
                                JoinOutcome.Blocked ->
                                    container.notify.error("You have been blocked from ${p.peek.name}.")
                                JoinOutcome.Failed ->
                                    container.notify.critical("Something went wrong while joining the group. Please try again.")
                            }
                        }
                    },
                )
            }

            Phase.PreviewError -> {
                LoadError(
                    onGoToDashboard = {
                        onDismiss()
                        nav.selectTab(social.betty.navigation.Tab.HOME)
                    },
                )
            }
        }
    }

    // Post-join / already-member confirm dialog — rendered outside the phase Column so it
    // survives after onDismiss() clears the sheet (the dialog state is remembered here).
    pendingConfirm?.let { (question, goThere) ->
        AlertDialog(
            onDismissRequest = { pendingConfirm = null },
            title = {
                Text(
                    text = question,
                    style = BettyTheme.type.body,
                    color = BettyTheme.colors.textPrimary,
                    modifier = Modifier.testTag("join-confirm"),
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    pendingConfirm = null
                    goThere()
                }) {
                    Text(
                        text = "GO THERE NOW",
                        style = BettyTheme.type.kicker,
                        color = Palette.orange,
                    )
                }
            },
            dismissButton = {
                TextButton(onClick = { pendingConfirm = null }) {
                    Text(
                        text = "STAY",
                        style = BettyTheme.type.kicker,
                        color = BettyTheme.colors.textMuted,
                    )
                }
            },
            containerColor = BettyTheme.colors.surface,
            titleContentColor = BettyTheme.colors.textPrimary,
        )
    }
}

// ---------------------------------------------------------------------------
// Invite preview (web JoinGroupModal)
// ---------------------------------------------------------------------------

@Composable
private fun InvitePreview(
    peek: GroupPeek,
    isJoining: Boolean,
    onDecline: () -> Unit,
    onAccept: () -> Unit,
) {
    val colors = BettyTheme.colors
    val type = BettyTheme.type

    Column {
        Column(
            modifier = Modifier
                .weight(1f, fill = false)
                .verticalScroll(rememberScrollState()),
        ) {
            // Header: 16:9 hero image with tournament icon overlay, or standalone round logo.
            val hasHeader = peek.headerImageUrl != null
            if (hasHeader) {
                HeroImage(
                    imageUrl = peek.headerImageUrl!!,
                    tournamentImageUrl = peek.tournamentImageUrl,
                )
            }

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(Space.l)
                    .padding(top = if (hasHeader) Space.m else Space.s),
                verticalArrangement = Arrangement.spacedBy(Space.xs),
            ) {
                // Large round tournament logo when no header image.
                if (!hasHeader && peek.tournamentImageUrl != null) {
                    RoundLogo(url = peek.tournamentImageUrl, size = 96.dp)
                    Spacer(Modifier.height(Space.s))
                }

                KickerText(text = "★ INVITED TO BET", color = Palette.orange)

                Text(
                    text = peek.name.uppercase(),
                    style = type.displayL,
                    color = colors.textPrimary,
                )

                if (!peek.tournamentName.isNullOrEmpty()) {
                    Text(
                        text = peek.tournamentName.uppercase(),
                        style = type.subhead,
                        color = colors.textSecondary,
                    )
                }

                // Description behind an orange left accent bar (components.md §4.3).
                // GroupPeek doesn't carry description — the field is absent in the wire model,
                // so this block is intentionally omitted. If the API is extended to include
                // description in GroupPeek, add it here via InsetPanel(accent = Palette.orange).

                Text(
                    text = "Lock in your bets every matchday, climb the standings, settle the banter.",
                    style = type.bodyRegular,
                    color = colors.textSecondary,
                )
            }
        }

        Divider(color = colors.overlay06)

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(Space.m),
            horizontalArrangement = Arrangement.spacedBy(Space.xs),
        ) {
            BettyButton(
                text = "NO THANKS",
                onClick = onDecline,
                variant = BettyButtonVariant.GHOST,
                modifier = Modifier
                    .weight(1f)
                    .testTag("join-decline"),
                block = true,
            )
            BettyButton(
                text = if (isJoining) "PLACING…" else "I'M IN →",
                onClick = onAccept,
                variant = BettyButtonVariant.PRIMARY,
                enabled = !isJoining,
                loading = isJoining,
                modifier = Modifier
                    .weight(1f)
                    .testTag("join-confirm"),
                block = true,
            )
        }
    }
}

// ---------------------------------------------------------------------------
// Hero image with optional tournament icon overlay
// ---------------------------------------------------------------------------

@Composable
private fun HeroImage(imageUrl: String, tournamentImageUrl: String?) {
    val colors = BettyTheme.colors
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .aspectRatio(16f / 9f),
    ) {
        AsyncImage(
            model = imageUrl,
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier.fillMaxSize(),
        )
        // Fallback tint while image loads.
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(colors.overlay04),
        )
        // Tournament logo overlaid bottom-start — offset downward so it bleeds into the
        // content below, matching the iOS `offset(y: 22)` treatment.
        if (tournamentImageUrl != null) {
            Box(
                modifier = Modifier
                    .align(Alignment.BottomStart)
                    .padding(start = Space.l)
                    .padding(bottom = (-22).dp),
            ) {
                RoundLogo(url = tournamentImageUrl, size = 56.dp)
            }
        }
    }
    // Compensate for the overlapping tournament logo.
    if (tournamentImageUrl != null) {
        Spacer(Modifier.height(22.dp))
    }
}

// ---------------------------------------------------------------------------
// Round logo (used both inline and as hero overlay)
// ---------------------------------------------------------------------------

@Composable
private fun RoundLogo(url: String, size: androidx.compose.ui.unit.Dp) {
    val colors = BettyTheme.colors
    AsyncImage(
        model = url,
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = Modifier
            .size(size)
            .clip(CircleShape)
            .background(colors.surface),
    )
}

// ---------------------------------------------------------------------------
// Preview fetch failure state
// ---------------------------------------------------------------------------

@Composable
private fun LoadError(onGoToDashboard: () -> Unit) {
    val colors = BettyTheme.colors
    val type = BettyTheme.type

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(Space.l)
            .testTag("join-error"),
        verticalArrangement = Arrangement.spacedBy(Space.m),
    ) {
        KickerText(text = "★ INVITED TO BET", color = Palette.orange)

        Text(
            text = "COULD NOT LOAD THIS INVITE",
            style = type.title2,
            color = colors.textPrimary,
        )

        Text(
            text = "The invite link may be invalid or expired. Please check the link and try again.",
            style = type.bodyRegular,
            color = colors.textSecondary,
        )

        BettyButton(
            text = "GO TO DASHBOARD",
            onClick = onGoToDashboard,
            variant = BettyButtonVariant.OUTLINE,
        )

        Spacer(Modifier.height(Space.m))
    }
}
